<?xml version="1.0"?>
<!-- Mikrorobot Mark I -->
<robot name="mikrorobot" xmlns:xacro="http://www.ros.org/wiki/xacro">

  <!-- Use this command to spawn the model in gazebo: rosrun gazebo_ros spawn_model -file `rospack find mikrorobot`/urdf/mikrorobot.urdf -urdf -z 1 -model mikrorobot -->

  <!-- Use this command to generate the urdf (use two -'s in front of inorder): rosrun xacro xacro -inorder mikrorobot.xacro > mikrorobot.urdf -->

  <!-- Constants for robot dimensions -->
  <xacro:property name="PI" value="3.1415926535897931"/>
  <xacro:property name="mass" value="10" /> <!-- arbitrary value for mass -->
  <xacro:property name="width" value="0.3" />
  <xacro:property name="depth" value="0.5" />
  <xacro:property name="height" value="0.27" />
  <xacro:property name="wheel_mass" value="1.5" />
  <xacro:property name="wheel_radius" value="0.1" />
  <xacro:property name="wheel_width" value="0.05" />
  <xacro:property name="camera_link" value="0.05" /> <!-- Size of square 'camera' box -->
  <xacro:property name="wheel_offset_x" value="0.184" /> <!-- Wheel offset depth-direction -->
  <xacro:property name="wheel_offset_y" value="0.198" /> <!-- Wheel offset width-direction -->
  <xacro:property name="wheel_offset_z" value="${-height/2 + 0.04}" /> <!-- Wheel offset height-direction -->

  <!-- Import all Gazebo-customization elements, including Gazebo colors -->
  <xacro:include filename="$(find mikrorobot)/urdf/mikrorobot.gazebo" />

    <!-- Base Link -->
  <link name="base_link">
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
	<box size="${depth} ${width} ${height}"/>
      </geometry>
    </collision>

    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
	<box size="${depth} ${width} ${height}"/>
      </geometry>
      <material name="yellow">
        <color rgba="0.8 0.8 0 1"/>
      </material>
    </visual>

    <inertial>
      <origin xyz="${depth/2} ${width/2} ${height/2}" rpy="0 0 0"/>
      <mass value="${mass}"/>
      <inertia
	  ixx="${mass / 12.0 * (width*width + depth*depth)}" ixy="0.0" ixz="0.0"
	  iyy="${mass / 12.0 * (height*height + depth*depth)}" iyz="0.0"
	  izz="${mass / 12.0 * (width*width + height*height)}"/>
    </inertial>
  </link>

  <!-- Front Right Wheel -->
  <joint name="front_right_wheel_joint" type="continuous">
    <parent link="base_link"/>
    <child link="front_right_wheel"/>
    <origin xyz="${wheel_offset_x} ${-wheel_offset_y} ${wheel_offset_z}" rpy="0 0 0" />
    <axis xyz="0 1 0" rpy="0 0 0" />
  </joint>

  <link name="front_right_wheel">
    <visual>
      <origin xyz="0 0 0" rpy="1.570795 0 0" />
      <geometry>
          <cylinder length="${wheel_width}" radius="${wheel_radius}" />
      </geometry>
      <material name="black">
        <color rgba="0.05 0.05 0.05 1"/>
      </material>
    </visual>

    <collision>
      <origin xyz="0 0 0" rpy="1.570795 0 0"/>
      <geometry>
          <cylinder length="${wheel_width}" radius="${wheel_radius}" />
      </geometry>
    </collision>

    <inertial>
      <mass value="${wheel_mass}" />
      <origin xyz="0 0 ${wheel_width/2}" rpy="0 0 0"/>
      <inertia ixx="${wheel_mass / 12.0*3* (wheel_radius * wheel_radius + wheel_width * wheel_width)}"
               ixy="0"
               ixz="0"
               iyy="${wheel_mass / 12.0*3* (wheel_radius * wheel_radius + wheel_width * wheel_width)}"
               iyz="0"
               izz="${wheel_mass * wheel_radius * wheel_radius / 2.0}" />
    </inertial>
  </link>

  <!-- Back Right Wheel -->
  <joint name="back_right_wheel_joint" type="continuous">
    <parent link="base_link"/>
    <child link="back_right_wheel"/>
    <origin xyz="${-wheel_offset_x} ${-wheel_offset_y} ${wheel_offset_z}" rpy="0 0 0" />
    <axis xyz="0 1 0" rpy="0 0 0" />
  </joint>

  <link name="back_right_wheel">
    <visual>
      <origin xyz="0 0 0" rpy="1.570795 0 0" />
      <geometry>
          <cylinder length="${wheel_width}" radius="${wheel_radius}" />
      </geometry>
      <material name="black"/>
    </visual>

    <collision>
      <origin xyz="0 0 0" rpy="1.570795 0 0"/>
      <geometry>
          <cylinder length="${wheel_width}" radius="${wheel_radius}" />
      </geometry>
    </collision>

    <inertial>
      <mass value="${wheel_mass}" />
      <origin xyz="0 0 ${wheel_width/2}" rpy="0 0 0"/>
      <inertia ixx="${wheel_mass / 12.0*3* (wheel_radius * wheel_radius + wheel_width * wheel_width)}"
               ixy="0"
               ixz="0"
               iyy="${wheel_mass / 12.0*3* (wheel_radius * wheel_radius + wheel_width * wheel_width)}"
               iyz="0"
               izz="${wheel_mass * wheel_radius * wheel_radius / 2.0}" />
    </inertial>
  </link>

  <!-- Front Left Wheel -->
  <joint name="front_left_wheel_joint" type="continuous">
    <parent link="base_link"/>
    <child link="front_left_wheel"/>
    <origin xyz="${wheel_offset_x} ${wheel_offset_y} ${wheel_offset_z}" rpy="0 0 0" />
    <axis xyz="0 1 0" rpy="0 0 0" />
  </joint>

  <link name="front_left_wheel">
    <visual>
      <origin xyz="0 0 0" rpy="1.570795 0 0" />
      <geometry>
          <cylinder length="${wheel_width}" radius="${wheel_radius}" />
      </geometry>
      <material name="black"/>
    </visual>

    <collision>
      <origin xyz="0 0 0" rpy="1.570795 0 0"/>
      <geometry>
          <cylinder length="${wheel_width}" radius="${wheel_radius}" />
      </geometry>
    </collision>

    <inertial>
      <mass value="${wheel_mass}" />
      <origin xyz="0 0 ${wheel_width/2}" rpy="0 0 0"/>
      <inertia ixx="${wheel_mass / 12.0*3* (wheel_radius * wheel_radius + wheel_width * wheel_width)}"
               ixy="0"
               ixz="0"
               iyy="${wheel_mass / 12.0*3* (wheel_radius * wheel_radius + wheel_width * wheel_width)}"
               iyz="0"
               izz="${wheel_mass * wheel_radius * wheel_radius / 2.0}" />
    </inertial>
  </link>

    <!-- Back Left Wheel -->
  <joint name="back_left_wheel_joint" type="continuous">
    <parent link="base_link"/>
    <child link="back_left_wheel"/>
     <origin xyz="${-wheel_offset_x} ${wheel_offset_y} ${wheel_offset_z}" rpy="0 0 0" />
    <axis xyz="0 1 0" rpy="0 0 0" />
  </joint>

  <link name="back_left_wheel">
    <visual>
      <origin xyz="0 0 0" rpy="1.570795 0 0" />
      <geometry>
          <cylinder length="${wheel_width}" radius="${wheel_radius}" />
      </geometry>
      <material name="black"/>
    </visual>

    <collision>
      <origin xyz="0 0 0" rpy="1.570795 0 0"/>
      <geometry>
          <cylinder length="${wheel_width}" radius="${wheel_radius}" />
      </geometry>
    </collision>

    <inertial>
      <mass value="${wheel_mass}" />
      <origin xyz="0 0 ${wheel_width/2}" rpy="0 0 0"/>
      <inertia ixx="${wheel_mass / 12.0*3* (wheel_radius * wheel_radius + wheel_width * wheel_width)}"
               ixy="0"
               ixz="0"
               iyy="${wheel_mass / 12.0*3* (wheel_radius * wheel_radius + wheel_width * wheel_width)}"
               iyz="0"
               izz="${wheel_mass * wheel_radius * wheel_radius / 2.0}" />
    </inertial>
  </link>

    <!-- Camera -->
  <link name="camera_link">

    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
	<cylinder length="0.03" radius ="0.02"/>
      </geometry>
    </collision>

    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
          <cylinder length="0.03" radius="0.02" />
      </geometry>
      <material name="black"/>
    </visual>

    <inertial>
      <mass value="1e-5" />
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <inertia ixx="1e-6" ixy="0" ixz="0" iyy="1e-6" iyz="0" izz="1e-6" />
    </inertial>
  </link>

  <joint name="camera_joint" type="fixed">
    <parent link="base_link"/>
    <child link="camera_link"/>
    <origin xyz="${depth/2 - 0.03} 0 ${height/2 + 0.02}" rpy="0 0 0" />
    <axis xyz="0 0 0" rpy="0 0 0" />
  </joint>

      <!-- Structure Camera -->
  <link name="camera_depth_optical_frame">

    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
	<cylinder length="0.03" radius ="0.02"/>
      </geometry>
    </collision>

    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
          <cylinder length="0.03" radius="0.02" />
      </geometry>
      <material name="black"/>
    </visual>

    <inertial>
      <mass value="1e-5" />
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <inertia ixx="1e-6" ixy="0" ixz="0" iyy="1e-6" iyz="0" izz="1e-6" />
    </inertial>
  </link>

  <joint name="structure_camera_joint" type="fixed">
    <parent link="base_link"/>
    <child link="camera_depth_optical_frame"/>
    <origin xyz="${depth/2 + 0.03} 0 ${height/2}" rpy="${-PI/2} 0 ${-PI/2}" />
    <axis xyz="0 0 0" rpy="0 0 0" />
  </joint>

      <!-- Laser -->
  <link name="laser_frame">

    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
	<cylinder length="0.03" radius ="0.02"/>
      </geometry>
    </collision>

    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
          <cylinder length="0.03" radius="0.02" />
      </geometry>
      <material name="black"/>
    </visual>

    <inertial>
      <mass value="1e-5" />
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <inertia ixx="1e-6" ixy="0" ixz="0" iyy="1e-6" iyz="0" izz="1e-6" />
    </inertial>
  </link>

  <joint name="laser_joint" type="fixed">
    <parent link="base_link"/>
    <child link="laser_frame"/>
    <origin xyz="${depth/2 - 0.05} 0 ${height/2 + 0.12}" rpy="0 0 ${PI}" />
    <axis xyz="0 0 0" rpy="0 0 0" />
  </joint>

        <!-- IMU -->
  <link name="imu_frame">

    <inertial>
      <mass value="1e-5" />
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <inertia ixx="1e-6" ixy="0" ixz="0" iyy="1e-6" iyz="0" izz="1e-6" />
    </inertial>
  </link>

  <joint name="imu_joint" type="fixed">
    <parent link="base_link"/>
    <child link="imu_frame"/>
    <origin xyz="${depth/2 - 0.10} ${-0.02} ${- height/2 + 0.4}" rpy="${PI/4} ${PI} 0" />
    <axis xyz="0 0 0" rpy="0 0 0" />
  </joint>

  <!-- Transmissions -->
  <transmission name="tran_f_r">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="front_right_wheel_joint">
      <hardwareInterface>EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor_f_r">
      <hardwareInterface>EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <transmission name="tran_b_r">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="back_right_wheel_joint">
      <hardwareInterface>EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor_b_r">
      <hardwareInterface>EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>


  <transmission name="tran_f_l">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="front_left_wheel_joint">
      <hardwareInterface>EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor_f_l">
      <hardwareInterface>EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <transmission name="tran_b_l">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="back_left_wheel_joint">
      <hardwareInterface>EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor_b_l">
      <hardwareInterface>EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

</robot>
